ios/LayoutReanimation/REASharedTransitionManager.m

```diff
@@ -0,0 +1,518 @@
+        continue;
+      }
+    }
+
+    REASnapshot *sourceViewSnapshot = [[REASnapshot alloc] init:viewSource withParent:viewSource.superview];
+    _snapshotRegistry[viewSource.reactTag] = sourceViewSnapshot;
+
+    REASnapshot *targetViewSnapshot;
+    if (addedNewScreen) {
+      targetViewSnapshot = [[REASnapshot alloc] init:viewTarget withParent:viewTarget.superview];
+      _snapshotRegistry[viewTarget.reactTag] = targetViewSnapshot;
+    } else {
+      targetViewSnapshot = _snapshotRegistry[viewTarget.reactTag];
+    }
+
+    if (!_currentSharedTransitionViews[viewSource.reactTag]) {
+      _currentSharedTransitionViews[viewSource.reactTag] = viewSource;
+    }
+    if (!_currentSharedTransitionViews[viewTarget.reactTag]) {
+      _currentSharedTransitionViews[viewTarget.reactTag] = viewTarget;
+    }
+
+    REASharedElement *sharedElement = [[REASharedElement alloc] initWithSourceView:viewSource
+                                                                sourceViewSnapshot:sourceViewSnapshot
+                                                                        targetView:viewTarget
+                                                                targetViewSnapshot:targetViewSnapshot];
+    [sharedElements addObject:sharedElement];
+  }
+  return sharedElements;
+}
+
+- (UIView *)getScreenForView:(UIView *)view
+{
+#if LOAD_SCREENS_HEADERS
+  UIView *screen = view;
+  while (![screen isKindOfClass:[RNSScreenView class]] && screen.superview != nil) {
+    screen = screen.superview;
+  }
+  if ([screen isKindOfClass:[RNSScreenView class]]) {
+    return screen;
+  }
+#endif
+  return nil;
+}
+
+- (UIView *)getStackForView:(UIView *)view
+{
+#if LOAD_SCREENS_HEADERS
+  if ([view isKindOfClass:[RNSScreenView class]]) {
+    if (view.reactSuperview != nil) {
+      if ([view.reactSuperview isKindOfClass:[RNSScreenStackView class]]) {
+        return view.reactSuperview;
+      }
+    }
+  }
+  while (view != nil && ![view isKindOfClass:[RNSScreenStackView class]] && view.superview != nil) {
+    view = view.superview;
+  }
+  if ([view isKindOfClass:[RNSScreenStackView class]]) {
+    return view;
+  }
+#endif
+  return nil;
+}
+
+- (void)observeChanges:(UIView *)view
+{
+  if (view == nil || [_screenHasObserver containsObject:view.reactTag]) {
+    return;
+  }
+  [_screenHasObserver addObject:view.reactTag];
+
+  [view addObserver:self forKeyPath:@"transitionDuration" options:NSKeyValueObservingOptionNew context:nil];
+  [view addObserver:self forKeyPath:@"activityState" options:NSKeyValueObservingOptionNew context:nil];
+
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    // it replaces method for RNSScreenView class, so it can be done only once
+    UIViewController *viewController = [view valueForKey:@"controller"];
+    [self swizzleMethod:@selector(viewDidLayoutSubviews)
+                   with:@selector(swizzled_viewDidLayoutSubviews)
+               forClass:[viewController class]];
```

Comment:
Passing `[viewController class]` and `[view class]` here feels risky to me - what if we pass a wrong view or someone subclasses `RNSScreenView` (for some reason) and we pass an object of that class here?
I think passing the classes by their name (i.e. `[RNSScreenView class]` or whatever else needs to be passed here) here would be a better idea, even if it makes it more dependent on Screens' implementation.

Reply:
Resolved in: https://github.com/software-mansion/react-native-reanimated/pull/4014/commits/e0f1b6a7325627667feeb895e7501f320bc797c5

Reply:
Resolved in: https://github.com/software-mansion/react-native-reanimated/pull/4014/commits/0608d3a2827d766efb0acb86f73b4a12b06b4b89